File: docs/文章/LCR084.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        helper(nums, 0, ans);
        return ans;
    }

    public void helper(int[] nums, int i, List<List<Integer>> ans) {
        if (i == nums.length) {
            List<Integer> temp = new ArrayList<>();
            for (int n : nums) {
                temp.add(n);
            }
            ans.add(temp);
        } else {
            HashSet<Integer> set =new HashSet<>();
            for (int j = i; j < nums.length; j++) {
                if (!set.contains(nums[j])) {
                    set.add(nums[j]);
                    
                    swap(nums, i, j);
                    helper(nums, i + 1, ans);
                    swap(nums, i, j);
                }
            }
        }
    }

    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 文章/LCR084.go
type HashSet map[int]bool

func permuteUnique(nums []int) [][]int {
    var ans [][]int
    helper(nums, 0, &ans)
    return ans
}

func helper(nums []int, i int, ans *[][]int) {
    if i == len(nums) {
        permutation := make([]int, len(nums))
        copy(permutation, nums)
        *ans = append(*ans, permutation)
    } else {
        set := make(HashSet)
        for j := i; j < len(nums); j++ {
            if !set[nums[j]] {
                set[nums[j]] = true
                swap(nums, i, j)
                helper(nums, i + 1, ans)
                swap(nums, i, j)
            }
        }
    }
}

func swap(nums []int, i, j int) {
    if i != j {
        nums[i], nums[j] = nums[j], nums[i]
    }
    
}
